Vault HA with Consul guide

yhyakuna opened the pull request:
***Vault HA with Consul*** guide is now ready for your review.  This was made possible based on Brian Shumate's doc --> https://support.hashicorp.com/hc/en-us/articles/360002183854

brianshumate commented on website/source/guides/operations/vault-ha-consul.html.md:
````diff
@@ -0,0 +1,621 @@
+```plaintext
+{
+  "server": true,
+  "node_id": "$NODE_ID",
````
This will need updating in the Consul configuration throughout as we'd later learned that it's `node_name` / `$NODE_NAME` we want here instead so this line should read:

```
"node_name": "$NODE_NAME",
```

yhyakuna replied:
Fixed. Thx!

brianshumate commented on website/source/guides/operations/vault-ha-consul.html.md:
```diff
@@ -0,0 +1,621 @@
+- `JOIN1**`, `**JOIN2`, `$JOIN3`: This example uses the `retry_join`
+method of joining the server agents to form a cluster; as such, the values for
+this guide would be `10.1.42.101`, `10.1.42.102`, and `10.1.42.103` respectively.
+- `$NODE_ID` this is a unique label for the node; in our case, this will be
```
This just needs updating to `$NODE_NAME` instead; the descriptive text should be good still.

brianshumate commented on website/source/guides/operations/vault-ha-consul.html.md:
```diff
@@ -0,0 +1,621 @@
+
+    {
+      "server": true,
+      "node_id": "consul_s1",
```
```
"node_name": "consul_s1",
```

instead here, please.

yhyakuna replied:
Fixed. Thx!

brianshumate commented on website/source/guides/operations/vault-ha-consul.html.md:
```diff
@@ -0,0 +1,621 @@
+
+    {
+      "server": true,
+      "node_id": "consul_s2",
```
```
"node_name": "consul_s2",
```

instead here, please.

yhyakuna replied:
Fixed. Thx!

brianshumate commented on website/source/guides/operations/vault-ha-consul.html.md:
```diff
@@ -0,0 +1,621 @@
+
+    {
+      "server": true,
+      "node_id": "consul_s3",
```
```
"node_name": "consul_s3",
```

instead here, please.

yhyakuna replied:
Fixed!

greenbrian commented on website/source/guides/operations/vault-ha-consul.html.md:
```diff
@@ -0,0 +1,621 @@
+  enough to inform your own production setup.
+---
+
+# Vault Haigh Availability (HA)
```
spelling

yhyakuna replied:
Good catch!  Thank you.

KFishner commented on website/source/guides/operations/vault-ha-consul.html.md:
```diff
@@ -0,0 +1,634 @@
+page_title: "Vault HA with Consul - Guides"
+sidebar_current: "guides-operations-vault-ha"
+description: |-
+  This guide This guide will walk you through a simple Vault Highly Available
```
remove `This guide`

yhyakuna replied:
Thx.  It's fixed now.

KFishner commented on website/source/guides/operations/vault-ha-consul.html.md:
```diff
@@ -0,0 +1,634 @@
+
+# Vault High Availability (HA)
+
+Vault can run in a high availability (HA) mode to protest against  outages by
```
s/protest/protect

yhyakuna replied:
Thx.  It's fixed now.

yhyakuna commented:
@KFishner   At the moment, _Vault Reference Architecture_ guide only has [one paragraph on HA](https://www.vaultproject.io/guides/operations/reference-architecture.html#high-availability), so I want to add a link saying something like, "***To learn more about setting up your Vault servers in HA mode, read Vault HA with Consul guide***".

KFishner commented:
@yhyakuna sounds good! That sentence could replace the current last line -  `For more information on this topic, please see the Vault site documentation about HA concepts.`

tdsacilowski commented:
I think we should add some more information about the fact that in a Vault cluster, that there's only one active node and x number of standbys... specifically around how standbys forward requests to the active node.

My understanding is that the values for:

api_addr (string: "") - Specifies the address (full URL) to advertise to other Vault servers in the cluster for client redirection. This value is also used for plugin backends. This can also be provided via the environment variable VAULT_API_ADDR.

cluster_addr (string: "") - – Specifies the address to advertise to other Vault servers in the cluster for request forwarding. This can also be provided via the environment variable VAULT_CLUSTER_ADDR.

Should ultimately resolve to the active node... whether they point to an external load balancer or Consul DNS (e.g. active.vault.service.consul).

I've gone into some detail on this in my HA & Replication Guide (https://docs.google.com/document/d/10XiFzneUP3ppDOg6e52Z_jDjM5TxU-EBicBtDk3SjwI/edit#heading=h.alspydkzxzmt)

... Just want to make sure we're consistent

yhyakuna commented:
@tdsacilowski  @brianshumate   I incorporated Teddy's feedback in the following manner:

- Added an additional sentence & a diagram in the intro section to talk mention **active** and **standby** mode. 
- Added **`api_addr`** in the Vault configuration files ([Step 4](https://github.com/hashicorp/vault/pull/4187/files#diff-feb4024abc5d32cd7f7700d9075f4233R479)) to teach the HA parameters

I kept it simple.  Let me know if you have further suggestion.

tdsacilowski commented:
Yoko... I'm going to retract my previous suggestions. Here are my thoughts...

There are two main patterns for talking to Vault, either directly, or through load balancers. In the case of the former, the _right_ way to set these is to be either the IP address or hostname of the node itself. For example:

vault_s1
api_addr = "https://10.1.42.201:8200"
cluster_addr = "https://10.1.42.201:8201"

vault_s2
api_addr = "https://10.1.42.202:8200"
cluster_addr = "https://10.1.42.202:8201"

Since we're keeping this guide simple, I'd just point to the direct connection, not using a load balancer (see https://github.com/hashicorp/vault/blob/56b5c8e8bdd78d43670ceee8511c95c9b08a231a/website/source/docs/concepts/ha.html.md#direct-access).

Additionally, when using Consul, Vault performs automatic host address detection so these values aren't _technically_ needed, but it would be a good idea to define them nonetheless.

yhyakuna commented:
@tdsacilowski   That's exactly what I did.

tdsacilowski commented:
> $API_ADDR: this should be set to the address which client (API) requests are to be redirected to. In our example, we want vault_s1 to be the active node, so this will be http://10.1.42.201:8200.

This can be the address of the node itself. Doesn't have to be the address of only the active node. See your config for vault_s2 (https://github.com/hashicorp/vault/blob/61f2092d78cfa2a6662972e8fa0941e65607b5fa/website/source/guides/operations/vault-ha-consul.html.md#vault_s2hcl-example).

If vault_s2 ever became the active node, it needs to advertise its own address, not the address of the other node.

In the scenario for this guide, each node should advertise it's own address.

yhyakuna commented:
@tdsacilowski  I think I got it right [this time](https://github.com/hashicorp/vault/pull/4187/files#diff-feb4024abc5d32cd7f7700d9075f4233R496).
